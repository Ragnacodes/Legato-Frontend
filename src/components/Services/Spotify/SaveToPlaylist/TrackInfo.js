import React from 'react';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';

export function TrackInfo({ track }) {
  return (
    <Card variant="outlined">
      <CardContent className="track-info">
        <img src={track.album.images[0].url} alt="track's album cover" />
        <div className="info">
          <div>
            <Typography className="name" variant="h6" component="h2">
              {track.name}
            </Typography>

            <Typography
              className="artists"
              variant="body2"
              color="textSecondary"
            >
              {track.artists.map((a) => a.name).join(', ')}
            </Typography>
          </div>
          <div>
            <Typography className="album" variant="body1">
              {track.album.name}
            </Typography>
          </div>
        </div>
      </CardContent>
    </Card>
  );
}

export default TrackInfo;
