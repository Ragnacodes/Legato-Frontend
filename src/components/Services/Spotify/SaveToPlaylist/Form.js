import React, { useState, useEffect } from 'react';
import { MenuItem, IconButton, TextField } from '@material-ui/core';
import { Refresh, Search } from '@material-ui/icons';
import { connect } from 'react-redux';
import { startGetConnections } from '../../../../actions/connections';
import * as actions from '../../../../actions/spotify';
import ServiceForm from '../../../PopoverForm';
import TrackInfo from './TrackInfo';
import { errorNotification } from '../../../Layout/Notification';
export function Form({
  id,
  data,
  track,
  connections,
  editElement,
  playlists,
  setAnchorEl,
  getPlaylists,
  getTrackInfo,
  setTrackInfo,
}) {
  const [info, setInfo] = useState({
    connection: data.connection || '',
    PlaylistId: data.PlaylistId || '',
    TrackId: data.TrackId || '',
    trackUrl: data.trackUrl || '',
    position: data.position || '',
  });

  const [errors, setErrors] = useState({});
  const [playlistLoading, setPlaylistLoading] = useState(true);
  const [trackLoading, setTrackLoading] = useState(false);

  useEffect(() => {
    getPlaylists().then(() => {
      setPlaylistLoading(false);
    });
  }, [getPlaylists]);

  const ExtractIdFromUrl = (url) => {
    try {
      const id = info['trackUrl'].split('spotify.com/track/')[1].split('?')[0];
      return id;
    } catch (err) {
      throw new Error('Please enter valid url.');
    }
  };

  const handleError = (err) => {
    if (err.message === 'Please enter valid url.') {
      setErrors((prev) => ({
        ...prev,
        trackUrl: err.message,
      }));
    } else {
      errorNotification(err.message);
    }
    setTrackInfo('');
  };

  const handleChange = (e) => {
    setInfo((prev) => ({
      ...prev,
      [e.target.name]: e.target.value,
    }));
  };

  const handleCancel = () => {
    setAnchorEl(null);
    setInfo({
      connection: data.connection || '',
      PlaylistId: data.PlaylistId || '',
      TrackId: data.TrackId || '',
      trackUrl: data.trackUrl || '',
      position: data.position || '',
    });
    setErrors({});
  };

  const handleSave = () => {
    try {
      const TrackId = ExtractIdFromUrl(info['trackUrl']);
      const updates = {
        name: info.name,
        data: { ...data, ...info, TrackId },
      };
      editElement(id, updates);
      setAnchorEl(null);
    } catch (err) {
      handleError(err);
    }
  };

  const searchTrack = () => {
    try {
      const TrackId = ExtractIdFromUrl(info['trackUrl']);
      getTrackInfo(TrackId)
        .then(() => {
          setTrackLoading(false);
          setInfo((p) => ({ ...p, TrackId }));
          setErrors((prev) => ({
            ...prev,
            trackUrl: '',
          }));
        })
        .catch((err) => {
          handleError(err);
        });
    } catch (err) {
      handleError(err);
    }
  };

  let disabledSave =
    !info['PlaylistId'] || !info['trackUrl'] || errors['trackUrl'];

  return (
    <ServiceForm
      className="save-to-playlist"
      title="Save to a Playlist"
      disabledSave={disabledSave}
      handleSave={handleSave}
      handleCancel={handleCancel}
    >
      <div className="playlist-field">
        {playlistLoading ? (
          <TextField
            className="text-field"
            size="small"
            label="Connection"
            value="Loading..."
            variant="outlined"
            disabled
          />
        ) : (
          <TextField
            name="PlaylistId"
            className="text-field"
            size="small"
            select
            label="Playlist"
            value={info['PlaylistId']}
            onChange={handleChange}
            variant="outlined"
          >
            {playlists.map((p) => (
              <MenuItem key={p.id} value={p.id}>
                {p.name}
              </MenuItem>
            ))}
          </TextField>
        )}

        <IconButton
          name="refreshPlaylists"
          size="small"
          className="add-icon"
          diabled={playlistLoading}
          onClick={() => {
            getPlaylists();
            setPlaylistLoading(true);
          }}
        >
          <Refresh />
        </IconButton>
      </div>

      <div className="track-field">
        <TextField
          name="trackUrl"
          className="text-field"
          label="Track URL"
          variant="outlined"
          size="small"
          value={info['trackUrl']}
          error={!!errors['trackUrl']}
          helperText={errors['trackUrl']}
          onChange={handleChange}
        />

        <IconButton
          name="searchForTrack"
          disabled={!info['trackUrl'] || trackLoading}
          size="small"
          className="add-icon"
          onClick={() => {
            searchTrack();
            setTrackLoading(true);
          }}
        >
          <Search />
        </IconButton>
      </div>

      {!!track && <TrackInfo track={track} />}
    </ServiceForm>
  );
}
const mapStateToProps = (state) => ({
  connections: state.connections,
  playlists: state.spotify.playlists,
  track: state.spotify.track,
});

const mapDispatchToProps = (dispatch) => ({
  getConnections: () => dispatch(startGetConnections()),
  getPlaylists: () => dispatch(actions.startGetPlaylists()),
  getTrackInfo: (track) => dispatch(actions.startGetTrackInfo(track)),
  setTrackInfo: (track) => dispatch(actions.setTrackInfo(track)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Form);
