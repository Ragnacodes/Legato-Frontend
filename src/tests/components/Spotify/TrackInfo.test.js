import React from 'react';
import { shallow } from 'enzyme';
import { TrackInfo } from '../../../components/Services/Spotify/SaveToPlaylist/TrackInfo';
import { Typography } from '@material-ui/core';

let wrapper, track;

beforeEach(() => {
  track = {
    name: 'track name',
    artists: [{ name: 'artist1' }, { name: 'artist2' }],
    album: {
      name: 'album name',
      images: [
        {
          url: 'image url',
        },
      ],
    },
  };
  wrapper = shallow(<TrackInfo track={track} />);
});

test('should render track info correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render track album cover', () => {
  expect(wrapper.find('img').length).toBe(1);
});

test('should render track name, artists, album', () => {
  expect(wrapper.find(Typography).length).toBe(3);
});

test('should render track name correctly', () => {
  const trackName = wrapper.find(Typography).find({ className: 'name' });
  expect(trackName.text()).toMatch(track.name);
});

test('should render track artists correctly', () => {
  const trackArtists = wrapper.find(Typography).find({ className: 'artists' });
  expect(trackArtists.text()).toMatch(
    `${track.artists[0].name}, ${track.artists[1].name}`
  );
});

test('should render track album correctly', () => {
  const trackAlbum = wrapper.find(Typography).find({ className: 'album' });
  expect(trackAlbum.text()).toMatch(track.album.name);
});

test('should render track album cover correctly', () => {
  const trackAlbumCover = wrapper.find('img');
  expect(trackAlbumCover.prop('src')).toMatch(track.album.images[0].url);
});
