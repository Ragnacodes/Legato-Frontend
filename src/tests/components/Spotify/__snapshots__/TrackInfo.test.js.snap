// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render track info correctly 1`] = `
<WithStyles(ForwardRef(Card))
  variant="outlined"
>
  <WithStyles(ForwardRef(CardContent))
    className="track-info"
  >
    <img
      alt="track's album cover"
      src="image url"
    />
    <div
      className="info"
    >
      <div>
        <WithStyles(ForwardRef(Typography))
          className="name"
          component="h2"
          variant="h6"
        >
          track name
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          className="artists"
          color="textSecondary"
          variant="body2"
        >
          artist1, artist2
        </WithStyles(ForwardRef(Typography))>
      </div>
      <div>
        <WithStyles(ForwardRef(Typography))
          className="album"
          variant="body1"
        >
          album name
        </WithStyles(ForwardRef(Typography))>
      </div>
    </div>
  </WithStyles(ForwardRef(CardContent))>
</WithStyles(ForwardRef(Card))>
`;
